controllers/release/release_adapter.go

```diff
@@ -228,24 +228,41 @@ func (a *Adapter) EnsureSnapshotEnvironmentBindingIsCreated() (results.Operation
 // EnsureTargetContextIsSet is an operation that will ensure that the targetContext of the adapter is not nil. It will
 // set it to a context from the workspace in the ReleasePlan if one is defined or to its own context if not.
 func (a *Adapter) EnsureTargetContextIsSet() (results.OperationResult, error) {
-	if a.targetContext == nil {
+	if a.targetContext != nil {
+		return results.ContinueProcessing()
+	}
+
+	// If the target namespace/workspace is not set we should get the target from the ReleasePlan
+	if a.release.Status.Target.Namespace == "" && a.release.Status.Target.Workspace == "" {
 		releasePlan, err := a.getReleasePlan()
 		if err != nil {
 			patch := client.MergeFrom(a.release.DeepCopy())
 			a.release.MarkInvalid(v1alpha1.ReleaseReasonReleasePlanValidationError, err.Error())
 			return results.RequeueOnErrorOrStop(a.client.Status().Patch(a.context, a.release, patch))
 		}
 
-		if releasePlan.Spec.Target.Workspace == "" {
-			a.logger.Info("Running in a traditional kubernetes environment")
-			a.targetContext = a.context
-		} else {
-			a.logger.Info("Running in a KCP environment")
-			a.targetContext = logicalcluster.WithCluster(a.context, logicalcluster.New(releasePlan.Spec.Target.Workspace))
+		patch := client.MergeFrom(a.release.DeepCopy())
+		if releasePlan.Spec.Target.Workspace != "" {
+			a.release.Status.Target.Workspace = releasePlan.Spec.Target.Workspace
+		}
+		if releasePlan.Spec.Target.Namespace != "" {
+			a.release.Status.Target.Namespace = releasePlan.Spec.Target.Namespace
 		}
+		err = a.client.Status().Patch(a.context, a.release, patch)
+		if err != nil {
+			return results.RequeueOnErrorOrStop(err)
+		}
+	}
 
-		a.syncer.SetContext(a.targetContext)
+	if a.release.Status.Target.Workspace == "" {
+		a.logger.Info("Running in a traditional kubernetes environment")
+		a.targetContext = a.context
+	} else {
+		a.logger.Info("Running in a KCP environment")
```

Comment:
By the way @johnbieren @davidmogar  I think this is not so correct as it gave me the impression that what is selecting if it is a KCP or not is the existance of the "Workspace" set in the Spec of the releasePlan. I guess it should check the existence of the cluster ID.

For instance, yesterday I forced adding "workspace" to the spec without an actual KCP environment, it breaks with _"Index with name cluster does not exist"_. 

So I guess by looking If the "logicalcluster" exists should be the proper way to define if it is a KCP env or not.

Reply:
Ok, good point. Checking the code we can have this at the top of the operation I think:
```
if _, ok := logicalcluster.ClusterFromContext(a.context); !ok {
    a.targetContext = a.context
    return results.ContinueProcessing()
}
// Actual code to set the targetContext for kcp using the Release status or the ReleasePlan
```
The cluster is set in the context in the `release_controller` for KCP, so it should work.

Reply:
I find the log line telling me if the controller started thinking we are in kcp or not *extremely* useful in my testing. Don't we also have some low priority task to add more logging to our controller? If it comes from main or somewhere else, I don't really care, but I would like to remain to be able to see what the controller thinks when it starts.

To Leandro's point, I think you *should* break if you define a target workspace that is not available to you. If I set my CRs up to release to a KCP workspace `foo` and `foo` doesn't exist to the controller, I don't want the controller to do some magic behind my back to still get releases going to who knows where. I want it to break and tell me that `foo` does not exist.

About removing the if check, I just had it for efficiency. There was no reason to go fetch CRs and parse them if the target context was already set

Reply:
@johnbieren @davidmogar I change the code to check if the Logical cluster exists but It comes with a problem. I can not test anything regarding KCP. It is possible to make it reach the workspace part for the code but then the Release Plan can not be found inside it due to the new context. Maybe there is a way to make it work, but could not do it so far.

I might change this back to the old way until we have a way to test the release service in KCP. The code is pushed anyway. I will move this to draft again to not bother with my commits and pushes.

Reply:
By the "old way" I meant the current way (checking if workspace parameter exists in the given spec) instead of my proposal of checking the existence of a logicalcluster and, thus, consider that it is a KCP env. My only point is if this is correct or not.

Reply:
As agreed on the meeting today:
1. The entry point to the function should be check if the Release status target is set to avoid loading the ReleasePlan.
2.  We are going to use logicalcluster so we can check if we are actually running on a KCP environment o not (nice suggestion @theflockers).
3. Based on the previous information we will validate if the workspace should be `""` (non-KCP) or have an actual value (KCP) and fail if necessary (abort other operations).
4. We'll be logging only whenever the Release status target gets updated.